crates/iota-genesis-builder/src/stardust/process_outputs.rs

```diff
@@ -67,9 +84,248 @@ pub fn is_participation_output(output: &Output) -> bool {
+    num_swapped_basic: u64,
+    num_swapped_timelocks: u64,
+    num_splits: u64,
+}
+
+impl<I> SwapSplitIterator<I> {
+    fn new(outputs: I, snapshot_timestamp_s: u32, swap_split_map: AddressSwapSplitMap) -> Self {
+        Self {
+            outputs,
+            snapshot_timestamp_s,
+            swap_split_map,
+            ordered_timelock_candidates: Default::default(),
+            split_basic_outputs: Default::default(),
+            num_swapped_basic: 0,
+            num_swapped_timelocks: 0,
+            num_splits: 0,
+        }
+    }
+
+    fn get_split_output(&mut self) -> Option<(OutputHeader, Output)> {
+        let (original_header, output) = self.split_basic_outputs.pop()?;
+        self.num_splits += 1;
+        let pos = self.split_basic_outputs.len();
+
+        let (transaction_id, output_index) = if original_header
+            .output_id()
+            .to_string()
+            .starts_with(VESTED_REWARD_ID_PREFIX)
+        {
+            // If the original basic output is a vesting output, generate the new OutputId
+            // as: original-transaction-id|index
+            // where index is a unique input
+            // index being a number in the range 1 to OUTPUT_INDEX_MAX is safe because
+            // vesting output indexes are always 0
+            // https://github.com/iotaledger/snapshot-tool-new-supply
+            let index = (pos as u16 % (OUTPUT_INDEX_MAX - 1)) + 1;
+            (
+                *original_header.output_id().transaction_id(),
+                OutputIndex::new(index).unwrap(),
+            )
+        } else {
+            // Else, generate the new OutputId as:
+            // DefaultHash("iota-genesis-outputs"|original-output-id|pos)|index
+            // where pos is a unique input
+            let index = pos as u16 % OUTPUT_INDEX_MAX;
+            let mut hasher = DefaultHash::default();
+            hasher.update(b"iota-genesis-outputs");
+            hasher.update(original_header.output_id().hash());
+            hasher.update(pos.to_le_bytes());
+            let hash = hasher.finalize();
+            (
+                TransactionId::new(hash.into()),
+                OutputIndex::new(index).unwrap(),
+            )
+        };
+
+        Some((
+            OutputHeader::new(
+                *transaction_id,
+                output_index,
+                *original_header.block_id(),
+                *original_header.ms_index(),
+                original_header.ms_timestamp(),
+            ),
+            output,
+        ))
+    }
+}
+
+impl<I> Iterator for SwapSplitIterator<I>
+where
+    I: Iterator<Item = Result<(OutputHeader, Output)>>,
+{
+    type Item = I::Item;
+
+    /// Get the next from the chained self.outputs iterator and apply the
+    /// SwapSplit filter if that's the case.
+    fn next(&mut self) -> Option<Self::Item> {
+        for mut output in self.outputs.by_ref() {
+            if let Ok((header, inner)) = &mut output {
```

Comment:
If this check fails, I think that means we will start looping below, and I think that's incorrect? I believe the `return Some(output);` line needs to be moved up one scope.

Reply:
Yeah, if that's the case then I think both should be changed